themes/qgrid/source/js/menu.js

```diff
@@ -15,41 +15,233 @@ function toggleVisibility(e) {
+		if (a && a.textContent) {
+			searchTarget = a;
+		}
+		if (highlightText(searchTarget, search) && !scrolled) {
+			scrolled = true;
+			searchTarget.scrollIntoView();
+		}
+	}
+}
+
+function tagsSearch(search) {
+	const nav = document.getElementById('nav');
+	const titlesOfMenuItems = nav.querySelectorAll('.title');
+	for (let titleOfMenuItem of titlesOfMenuItems) {
+		const menuItem = titleOfMenuItem.parentElement;
+		const tags = menuItem.dataset.tag || '';
+		for (let tag of tags.split(',')) {
+			if (tagFormatting(menuItem, tag, search)) {
+				showTag(menuItem, tag, search);
+				const dataShowTag = menuItem.getElementsByTagName('data-show-tag')[0].querySelector('a');
+				highlightText(dataShowTag, search);
+				break;
+			}
+		}
+	}
+}
+
+function menuItemsSearch(search) {
+	const nav = document.getElementById('nav');
+	const titlesOfMenuItems = nav.querySelectorAll('.title');
+	for (let titleOfMenuItem of titlesOfMenuItems) {
+		highlightText(titleOfMenuItem, search);
+	}
+}
+
+function tagFormatting(menuItem, tag, search) {
+	const titleOfMenuItem = menuItem.querySelector('.title');
+	if (tag.toLowerCase().indexOf(search) >= 0 && search != '') {
+		menuItem.style.display = '';
+		titleOfMenuItem.classList.add('menu-item');
+		if (titleOfMenuItem.textContent[0] != '/') {
+			titleOfMenuItem.prepend("/ ");
+		}
+		return true;
+	}
+	else {
+		if (search != '') {
+			menuItem.style.display = 'none';
+		}
+		else {
+			menuItem.style.display = '';
+		}
+		const dataShowTag = menuItem.getElementsByTagName('data-show-tag')[0];
+		if (dataShowTag) {
+			dataShowTag.style.display = 'none';
+		}
+		titleOfMenuItem.classList.remove('menu-item');
+		titleOfMenuItem.textContent = titleOfMenuItem.textContent.replace("/ ", '');
+		return false;
+	}
+}
+
+function showTag(menuItem, tag, search) {
+	for (let searchTarget of tag.split(' ')) {
+		let indexTarget = -1;
+		for (let arraySearch of search.split(' ')) {
+			if (indexTarget < 0 && arraySearch != '') {
+				indexTarget = searchTarget.toLowerCase().indexOf(arraySearch);
+			}
+		}
+		const indexSearch = search.split(' ').indexOf(searchTarget.toLowerCase());
+		if ((indexTarget >= 0 || indexSearch >= 0) && search != '') {
+			if (menuItem.getElementsByTagName('data-show-tag').length === 0) {
+				createDataShowTag(menuItem);
+			}
+			const dataShowTag = menuItem.getElementsByTagName('data-show-tag')[0];
+			dataShowTag.style.display = '';
+			const a = dataShowTag.getElementsByTagName('a')[0];
+			a.textContent = getFoundTagsContent(tag, searchTarget);
+			a.setAttribute('href', menuItem.querySelector('.title').getAttribute('href'));
+			a.style.display = 'inline';
+		}
+	}
+}
+
+function createDataShowTag(menuItem) {
+	const newTag = document.createElement('data-show-tag');
+	newTag.appendChild(document.createElement('a'));
+	menuItem.insertBefore(newTag, menuItem.children[0]);
+}
+
+function highlightText(foundElement, search) {
+	const { textContent } = foundElement;
```

Comment:
please investigate this code: https://github.com/qgrid/ng2/blob/master/src/lib/pipes/highlight.pipe.ts
it's much simpler to read than existing one.